citesphere/src/main/java/edu/asu/diging/citesphere/core/bib/impl/CitationVersionDao.java

```diff
@@ -0,0 +1,81 @@
+package edu.asu.diging.citesphere.core.bib.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.javers.core.Javers;
+import org.javers.core.metamodel.object.CdoSnapshot;
+import org.javers.core.metamodel.object.SnapshotType;
+import org.javers.repository.jql.QueryBuilder;
+import org.javers.shadow.Shadow;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import edu.asu.diging.citesphere.core.bib.ICitationVersionsDao;
+import edu.asu.diging.citesphere.core.model.bib.CitationVersion;
+import edu.asu.diging.citesphere.model.bib.ICitation;
+import edu.asu.diging.citesphere.model.bib.impl.Citation;
+
+@Repository
+public class CitationVersionDao implements ICitationVersionsDao {
+
+    @Autowired
+    private Javers javers;
+
+    private static final String GROUP_PROPERTY = "group";
+
+    @Override
+    public List<CitationVersion> getVersions(String groupId, String key, int page, int pageSize) {
+        int offset = page * pageSize;
+        List<CitationVersion> result = new ArrayList<>();
+        List<CdoSnapshot> versions = javers.findSnapshots(QueryBuilder.byInstanceId(key, Citation.class)
+                .withSnapshotTypeUpdate().skip(offset).limit(pageSize).build());
+        result.addAll(versions.stream().filter(version -> version.getPropertyValue(GROUP_PROPERTY).equals(groupId))
+                .map(version -> toCitationVersion(version, key)).collect(Collectors.toList()));
```

Comment:
see my comment on the plugin, what about `withCommitProperty`?

Reply:
also, why don't you use the list returned by `collect`? why adding it to a new list?

Reply:
> see my comment on the plugin, what about `withCommitProperty`?

`withCommitProperty` refers to the properties of the commit metadata. It's empty as of now. I had used it earlier to manually add the citation key and then use `withCommitProperty` to extract all the versions. I can add the group Id which would be better than the current logic.

Reply:
yes, that sounds like the better approach.